### PR TITLE
Pushgateway metrics

### PR DESCRIPTION
**Pull request by jespada**
This PR adds support for [Pushgateway](https://github.com/prometheus/pushgateway).

If there are no config entries for TLS in pushgw it will use plain http.

### CODE REVIEW

**Comment by mcorbin** [comment]
> If there are no config entries for TLS in pushgw it will use plain http.

It now supports HTTPS right ?

**Comment by jespada** [comment]
> > If there are no config entries for TLS in pushgw it will use plain http.
> 
> It now supports HTTPS right ?

@mcorbin yes, you can use it like so: https://gist.github.com/jespada/a4b48f7348c2f7576406933f8334b40f

File: pushgw/root.go [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+		if err != nil {
+			return nil, errors.Wrap(err, "can't load certificate")
+		}
+
+		caCertPool := x509.NewCertPool()
+		caCertPool.AppendCertsFromPEM(caCert)
+
+		tlsConfig = &tls.Config{
+			Certificates: []tls.Certificate{cert},
+			RootCAs:      caCertPool,
```
Comment (jespada) [inline pushgw/root.go]: Should we provide config option for `InsecureSkipVerify` ?

**Comment by falzm** [comment]
Hi @jespada, thank you for your contribution however IMO the support for the Prometheus Push Gateway should be implemented as a metric target similar to `metrics.PrometheusConfiguration`, `metrics.CollectdConfiguration` etc  – not as a new top-level component.

**Comment by jespada** [comment]
> Hi @jespada, thank you for your contribution however IMO the support for the Prometheus Push Gateway should be implemented as a metric target similar to `metrics.PrometheusConfiguration`, `metrics.CollectdConfiguration` etc – not as a new top-level component.

Hey @falzm , well I'm not sure, that was the initial idea but after a quick discussion with @mcorbin we decided that was best to do the way is done on this PR. I don't have a strong opinion on this, so its up to you both.

**Comment by falzm** [comment]
@jespada @mcorbin I really feel that the direction that this implementation is heading to is not a correct one: the Prometheus Push Gateway is just another metrics exporter, and even if it slightly differs from the others in that it's a client and not a server I'd prefer to align its implementation on the others to ensure consistency across the whole library.

Let me know if you prefer me to take over this task if it's more effort than you're willing to put, it's no problem for me :)

**Comment by jespada** [comment]
> @jespada @mcorbin I really feel that the direction that this implementation is heading to is not a correct one: the Prometheus Push Gateway is just another metrics exporter, and even if it slightly differs from the others in that it's a client and not a server I'd prefer to align its implementation on the others to ensure consistency across the whole library.
> 
> Let me know if you prefer me to take over this task if it's more effort than you're willing to put, it's no problem for me :)

Thanks for your feedback @falzm  I'll work on it this week at some point.

**Comment by falzm** [comment]
Adding the WIP label pending PR update.

**Comment by jespada** [comment]
Closing this one in favour of https://github.com/exoscale/go-reporter/pull/11

**Pull request closed without merging**